src/haz3lmenhir/Parser.mly

```diff
@@ -0,0 +1,269 @@
+%token L_AND
+%token L_OR
+%token L_NOT
+(*bitwise ops*)
+%token COMMA
+%token COLON
+%token EOF
+%token IN
+%token UNIT
+%token END
+
+(* type tokens *)
+%token INT_TYPE
+%token FLOAT_TYPE
+%token BOOL_TYPE
+%token STRING_TYPE
+%token UNKNOWN
+%token INTERNAL
+
+%token IF
+%token THEN
+%token ELSE
+
+%token SEMI_COLON
+
+
+
+(* Precedences *)
+
+%nonassoc IF_EXP
+%nonassoc LET_EXP
+
+%left DASH_ARROW
+
+%left PLUS MINUS TIMES POWER DIVIDE DOUBLE_EQUAL NOT_EQUAL LESS_THAN_EQUAL GREATER_THAN_EQUAL 
+
+%left GREATER_THAN LESS_THAN
+
+%left PLUS_FLOAT MINUS_FLOAT TIMES_FLOAT POWER_FLOAT DIVIDE_FLOAT DOUBLE_EQUAL_FLOAT NOT_EQUAL_FLOAT LESS_THAN_FLOAT LESS_THAN_EQUAL_FLOAT GREATER_THAN_FLOAT GREATER_THAN_EQUAL_FLOAT
+
+(* Other *)
+%left CONS
+%left OPEN_PAREN
+%left QUESTION
+%left COLON
+%left COMMA
+%left AT_SYMBOL
+%left SEMI_COLON
+%left IN
+%left DOLLAR_SIGN
+%left L_NOT L_AND L_OR
+
+(* Might not be correct - milan *)
+%left IDENT
+
+%type <AST.exp> exp
+
+%start <AST.exp> program
+
+%%
+
+program:
+    | e = exp; EOF {e}
+
+%inline intOp:
+    | PLUS { IntOp(Plus) }
+    | MINUS { IntOp(Minus) }
+    | TIMES { IntOp(Times) }
+    | POWER { IntOp(Power) }
+    | DIVIDE { IntOp(Divide) }
+    | DOUBLE_EQUAL { IntOp(Equals) }
+    | NOT_EQUAL { IntOp(NotEquals) }
+    (* | LESS_THAN { IntOp(LessThan) } *)
+    | LESS_THAN_EQUAL { IntOp(LessThanOrEqual) }
+    | GREATER_THAN { IntOp(GreaterThan) }
+    | GREATER_THAN_EQUAL { IntOp(GreaterThanOrEqual) }
+
+
+%inline floatOp:
+    | PLUS_FLOAT { FloatOp(Plus) }
+    | MINUS_FLOAT { FloatOp(Minus) }
+    | TIMES_FLOAT { FloatOp(Times) }
+    | POWER_FLOAT { FloatOp(Power) }
+    | DIVIDE_FLOAT { FloatOp(Divide) }
+    | DOUBLE_EQUAL_FLOAT { FloatOp(Equals) }
+    | NOT_EQUAL_FLOAT { FloatOp(NotEquals) }
+    | LESS_THAN_FLOAT { FloatOp(LessThan) }
+    | LESS_THAN_EQUAL_FLOAT { FloatOp(LessThanOrEqual) }
+    | GREATER_THAN_FLOAT { FloatOp(GreaterThan) }
+    | GREATER_THAN_EQUAL_FLOAT { FloatOp(GreaterThanOrEqual) }
+
+%inline boolOp:
+    | L_AND { BoolOp(And) }
+    | L_OR { BoolOp(Or) }
+
+%inline binOp:
+    | i = intOp { i }
+    | f = floatOp { f }
+    | b = boolOp { b }
+
+binExp:
+    | e1 = exp; b = binOp; e2 = exp { BinExp (e1, b, e2) }
+
+typ:
+    | QUESTION; s = STRING { InvalidTyp(s) }
+    | INT_TYPE { IntType }
+    | FLOAT_TYPE { FloatType }
+    | BOOL_TYPE { BoolType }
+    | STRING_TYPE { StringType }
+    | UNKNOWN; INTERNAL { UnknownType(Internal) }
+    | UNIT { UnitType }
+    | OPEN_PAREN; types = separated_list(COMMA, typ); CLOSE_PAREN { TupleType(types) }
+    | OPEN_SQUARE_BRACKET; t = typ; CLOSE_SQUARE_BRACKET { ArrayType(t) }
+    | t1 = typ; DASH_ARROW; t2 = typ { ArrowType(t1, t2) }
+
+pat:
+    | QUESTION; s = STRING { InvalidPat(s) }
+    | WILD { WildPat }
+    | QUESTION { EmptyHolePat }
+    | OPEN_SQUARE_BRACKET; l = separated_list(COMMA, pat); CLOSE_SQUARE_BRACKET; { ListPat(l) }
+    | c = CONSTRUCTOR_IDENT; COLON; t = typ;  { ConstructorPat(c, t) }
```

Comment:
It's worth noting maketerm always parses the constructor call as unknown https://github.com/hazelgrove/hazel/blob/dev/src/haz3lcore/statics/MakeTerm.re#L193. I guess it's useful to have this with the type annotation for tests but it's incompatible syntax with the main editor.

Reply:
I think I miscommunicated what I was saying with the cast/constructor type. I pushed a new commit https://github.com/hazelgrove/hazel/pull/1220/commits/ac4e9f9267679973f10bf14bdc3b639bee2ed499 that does what I was describing. Where `A : T` is the unknown constructor A cast to T and `A ~ T` is the constructor A of type `T`

Reply:
Ah ok, got it. Thanks!